file.go

```diff
@@ -114,121 +113,105 @@ func (obj *FileRes) Watch(processChan chan Event) {
+	var watching = false
 
 	for {
-		current = strings.Join(patharray[0:index], "/")
-		if current == "" { // the empty string top is the root dir ("/")
-			current = "/"
-		}
+		watchPath = "/" + strings.Join(patharray[0:watchDepth], "/")
+
 		if DEBUG {
-			log.Printf("File[%v]: Watching: %v", obj.GetName(), current) // attempting to watch...
+			log.Printf("File[%v]: Watching: %v", obj.GetName(), watchPath) // attempting to watch...
 		}
-		// initialize in the loop so that we can reset on rm-ed handles
-		err = watcher.Add(current)
-		if err != nil {
-			if DEBUG {
-				log.Printf("File[%v]: watcher.Add(%v): Error: %v", obj.GetName(), current, err)
-			}
-			if err == syscall.ENOENT {
-				index-- // usually not found, move up one dir
-			} else if err == syscall.ENOSPC {
-				// XXX: occasionally: no space left on device,
-				// XXX: probably due to lack of inotify watches
-				log.Printf("%v[%v]: Out of inotify watches!", obj.Kind(), obj.GetName())
-				log.Fatal(err)
-			} else {
-				log.Printf("Unknown file[%v] error:", obj.Name)
+
+		if !watching {
+			err = watcher.Add(watchPath)
+			if err != nil {
+				if DEBUG {
+					log.Printf("File[%v]: watcher.Add(%v): Error: %v", obj.GetName(), watchPath, err)
+				}
+				if err == syscall.ENOENT {
+					watchDepth-- // usually not found, move up one dir
+					if watchDepth < 0 {
+						log.Fatal("somehow trying to watch file above the fs root")
+					}
+					continue
+				}
+				if err == syscall.ENOSPC {
+					// XXX: occasionally: no space left on device,
+					// XXX: probably due to lack of inotify watches
+					log.Printf("%v[%v]: Out of inotify watches!", obj.Kind(), obj.GetName())
+				} else {
+					log.Printf("Unknown file[%v] error:", obj.Name)
+				}
 				log.Fatal(err)
 			}
-			index = int(math.Max(1, float64(index)))
-			continue
+			watching = true
 		}
 
 		obj.SetState(resStateWatching) // reset
 		select {
 		case event := <-watcher.Events:
 			if DEBUG {
-				log.Printf("File[%v]: Watch(%v), Event(%v): %v", obj.GetName(), current, event.Name, event.Op)
+				log.Printf("File[%v]: Watch(%v), Event(%v): %v", obj.GetName(), watchPath, event.Name, event.Op)
 			}
-			cuuid.SetConverged(false) // XXX: technically i can detect if the event is erroneous or not first
-			// the deeper you go, the bigger the deltaDepth is...
-			// this is the difference between what we're watching,
-			// and the event... doesn't mean we can't watch deeper
-			if current == event.Name {
-				deltaDepth = 0 // i was watching what i was looking for
-
-			} else if HasPathPrefix(event.Name, current) {
-				deltaDepth = len(PathSplit(current)) - len(PathSplit(event.Name)) // -1 or less
-
-			} else if HasPathPrefix(current, event.Name) {
-				deltaDepth = len(PathSplit(event.Name)) - len(PathSplit(current)) // +1 or more
 
-			} else {
+			if !HasPathPrefix(event.Name, cleanObjPath) && !HasPathPrefix(cleanObjPath, event.Name) {
 				// TODO different watchers get each others events!
```

Comment:
I think the correct behaviour of the library is to only give you events you've asked for, but in the scenario that it gives us events we don't want (as seen here) we just ignore them.